### DISCUSSION TITLE
Making a new SpaCy compatible model from scratch without Thinc. Possible?

### DISCUSSION THREAD

**Discussion by koaning** [explosion/spaCy#6948] [Help: Coding & Implementations] [answered]
I'm interested in making scikit-learn/vowpal wabbit available in spaCy through a community package. I figured it'd be a fun exercise to start to understand the internals of spaCy better but it might also give me a meaningful tool to measure/understand when simple models get beaten by more complex models offered by spaCy. 

The thing is though ... these systems that I'm considering don't use Thinc and they certainly don't expose a Thinc Model. Looking at the [docs for the `TrainablePipe` component](https://spacy.io/usage/processing-pipelines#trainable-components) it says that it requires a Thinc Model. So I might need to consider another route. 

There's a simple way to add a scikit-learn model to spaCy, simply as a `Language.component`. I could add a model by doing something like this from Jupyter; 

```python
import spacy
import pandas as pd
from spacy.language import Language

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline, make_union


df = pd.read_json("data/intents.jsonl", lines=True)
pipe = make_pipeline(CountVectorizer(), LogisticRegression())

pipe.fit(df['text'], df['label'])

nlp = spacy.load("en_core_web_md")

@Language.component("sklearn-cat")
def my_component(doc):
    pred = pipe.predict([doc.text])[0]
    proba = pipe.predict_proba([doc.text]).max()
    doc.cats[pred] = proba
    return doc

nlp.add_pipe("sklearn-cat")
nlp("can you flip a coin?").cats
# {'flip_coin': 0.9747356912446946}
```

This would technically work in the sense that I can use the output from sklearn in spaCy. But it's a bit hacky for the long term. The main downside is that I'm not using the config system to define my scikit-learn model which means that I would need two systems to use while benchmarking. You would also need to serialize the model separately. So this path doesn't 100% feel right either. 

So I'm wondering. Does anybody have any pointers on what direction might be best here?

**Comment by honnibal** [comment, marked as answer]
The most straight-forward solution is probably to subclass `Pipe` rather than `TrainablePipe` (which does make me thing `TrainablePipe` might not have been the best name 🤔 ). 

A more ideal solution would be to add a Vowpal Wabbit shim for Thinc, which would mean you wouldn't need to mess with the `TextCategorizer` component at all. I think this should work fine because VW supports online learning.

I don't see an elegant way to have a shim for scikit-learn models though, because they have an API that doesn't match up well to Thinc. They train via a single `.fit()` method, rather than with updates per batch.

**Reply by koaning** [reply]
Scikit-Learn does have support for this actually, but not all models have this feature. Some feature transformers and models implement a `.fit_partial(X, y)` method which allows you to pass in data batch by batch. 

In particular, I'm looking at [the hashing trick](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html#sklearn.feature_extraction.FeatureHasher) to generate sparse features. These can then be used by [Naive Bayes](https://scikit-learn.org/0.15/modules/generated/sklearn.naive_bayes.MultinomialNB.html#sklearn.naive_bayes.MultinomialNB), [PA Classifier](https://scikit-learn.org/0.15/modules/generated/sklearn.linear_model.PassiveAggressiveClassifier.html#sklearn.linear_model.PassiveAggressiveClassifier) or [SGD Classifier](https://scikit-learn.org/0.15/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier). These all accept sparse features and have `fit_partial` implemented.

**Reply by koaning** [reply]
One thing about the `Pipe` vs `TrainablePipe`. It might be good to add `Pipe` to the API documentation. I could only find `TrainablePipe` [here](https://spacy.io/api/pipe) which led me to believe it was the intended component I use.

**Reply by ines** [reply]
(I'm not 100% sure I agree with Matt re `Pipe` – it's mostly internals and defines the general interface to use for pipeline components. The main reason it's currently not documented is that we _don't_ want users to think that a pipeline component needs to inherit from `Pipe`.)

Looking at your example, if you wanted to implement this in a fully standalone way, you'd probably want a [stateful component](https://spacy.io/usage/processing-pipelines#custom-components-factories) that holds your sklearn model and implements all the methods called by spaCy (see [`TrainablePipe`](https://spacy.io/api/pipe) for how those methods should work): 

* `update`: update your model with a batch of `Example` objects
* `predict` predict whatever you want to predict for a batch of `Doc` objects
* `set_annotations`: given what you predicted, store the annotations on the `Doc`
* `__call__`: runs when component is called on a `Doc`, predict and set annotations
* `score`: compute scores for your component
* `to_disk`/`from_disk`/`to_bytes`/`from_bytes`: Save/load component data to/from disk/bytes

If you want to make this all elegant and composable, your component could take the sklearn model as an argument, generated by a function in the function registry. This would make it easy to swap out different architectures via the config.

```python
@Language.factory("sklearn-cat")
def make_sklearn_cat(nlp, name, sklearn_model):
    return SklearnCat(nlp, name, sklearn_model)

class SklearnCat:
    def __init__(self, nlp, name, sklearn_model):
        self.nlp = nlp
        self.name = name
        self.sklearn_model = sklearn_model

    # etc.

@spacy.registry.architectures("sklearn_cat_basic.v1")
def make_sklearn_cat_basic():
    return make_pipeline(CountVectorizer(), LogisticRegression())
```

Now if you want to try out a different sklearn pipeline, you'd just register another architecture. And if you want your architecture to be configurable with settings, just add an argument to the function and provide its value via the config.

In the config, the above would look like this:

```ini
[components.textcat]
factory = "sklearn-cat"

[components.textcat.sklearn_model]
@architectures  = "sklearn_cat_basic.v1"
```

That said, it'd definitely be great to have a Shim/Wrapper in Thinc for Scikit-learn models that support updates per batch, just like we currently do for PyTorch and TensorFlow. The advantage here is that you'll need to write much less boilerplate specific to your model: if it's an instance of `Model`, spaCy knows how to talk to it, it serializes, exposes the same methods as all other architectures, and so on. And it can easily slot into any other component – or, if you write your custom text classification component and want to test it with a PyTorch implementation, just swap in the wrapped model and you won't have to change a single line of your component code.

**Reply by koaning** [reply]
Thanks, @ines! This helps. I'll play around with this here. I wrote some extensions for scikit-learn such that we have a proper `PartialPipeline` variant of the `Pipeline` so that we can handle the `partial_fit` API. 

I'll try to get this working first for spaCy directly. Once I get that working and get a feel for how things connect some more I'll explore the `Shim` further.

**Reply by koaning** [reply]
I think I've got the ball rolling. This now works in my notebook; 

```python
nlp = spacy.load("en_core_web_sm")
conf = {"sklearn_model": "@sklearn_model_basic.v1", "label": "pos", "classes": ["pos", "neg"]}
nlp.add_pipe("sklearn-cat", config=conf)

texts = ["you are a nice person", "this is a great movie", "i do not like cofee"]
labels = ["pos", "pos", "neg"]

with nlp.select_pipes(enable="sklearn-cat"):
    optimizer = nlp.resume_training()
    for itn in range(100):
        for t, lab in zip(texts, labels):
            doc = nlp.make_doc(t)
            example = Example.from_dict(doc, {"cats": {"pos": lab}})
            nlp.update([example], sgd=optimizer)
            
nlp("you are a nice person").cats # {'pos': 0.9979167909733176}
nlp("coffee i do not like").cats # {'neg': 0.990049724779963}
```

There's a work in progress PR [here on my own package repo](https://github.com/koaning/tokenwiser/pull/27/files?file-filters%5B%5D=.py#r571986483). There's still some work to do on the scikit-learn side of things. I still need to confirm that everything works neatly outside of a notebook (haven't implemented the serialization stuff yet) but it's a place to start 😄 .

**Comment by ines** [comment]
> The thing is though ... these systems that I'm considering don't use Thinc and they certainly don't expose a Thinc Model. 

That's pretty normal and expected, though – it might be worth mentioning here that a Thinc's [`Model`](http://thinc.ai/docs/api-model) mostly exists to give the model implementations powering the pipeline components a unified API that spaCy can interact with. So the idea is to wrap your model in an object that spaCy knows how to talk to.

For PyTorch, TensorFlow and MXNet we have [simple wrappers](https://thinc.ai/docs/usage-frameworks) that return a `Model` instance and make it easy to plug implementations in different frameworks. This means that the spaCy component doesn't have to care about whether it's model architecture is implemented in PyTorch or TF, or something completely different. And you can even mix and match models and layers implemented in different frameworks, or have different components be powered by different frameworks, without having to change the component itself.

**Reply by koaning** [reply]
I've never really used `Thinc` before so I wasn't aware of the `Shim`. Thanks for the link!

**Comment by koaning** [comment]
To anybody interested, I've started working on a package for this. It's called tokenwiser (it's a bag of tricks, not words!). It can be found [here](https://koaning.github.io/tokenwiser/index.html).
